File: android/src/main/kotlin/dev/scankit/scankit/ScankitPlugin.kt
/**
 * ScankitPlugin.kt
 * ScanKit - Modern barcode scanner for Flutter
 *
 * Main plugin entry point that implements the Pigeon-generated ScannerHostApi.
 * This file coordinates between Flutter and native scanner implementations.
 *
 * Architecture:
 * - ScankitPlugin: Main plugin, handles Pigeon API calls
 * - EmbeddedScannerView: Platform view for embedded scanning
 * - ScannerActivity: Full-screen scanner activity
 * - ImageScanner: Static image scanning (scanner/)
 * - BarcodeFormatMapper: Format string <-> ML Kit constant mapping (utils/)
 */
package dev.scankit.scankit

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.hardware.camera2.CameraCharacteristics
import android.hardware.camera2.CameraManager
import android.provider.MediaStore
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.IntentSenderRequest
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import com.google.mlkit.vision.barcode.BarcodeScanning
import com.google.mlkit.vision.barcode.BarcodeScannerOptions
import com.google.mlkit.vision.barcode.common.Barcode
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.documentscanner.GmsDocumentScanner
import com.google.mlkit.vision.documentscanner.GmsDocumentScannerOptions
import com.google.mlkit.vision.documentscanner.GmsDocumentScanning
import com.google.mlkit.vision.documentscanner.GmsDocumentScanningResult
import dev.scankit.scankit.scanner.ImageScanner
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

/**
 * Main ScanKit Flutter plugin class.
 *
 * Implements:
 * - FlutterPlugin: For Flutter engine attachment
 * - ActivityAware: For activity lifecycle management
 * - ScannerHostApi: Pigeon-generated API interface
 */
class ScankitPlugin : FlutterPlugin, ActivityAware, ScannerHostApi {

    // MARK: - Properties

    private var activity: Activity? = null
    private var context: Context? = null
    private var activityBinding: ActivityPluginBinding? = null
    private var flutterApi: ScannerFlutterApi? = null

    // Pending callbacks for async operations
    private var pendingCallback: ((Result<BarcodeResult?>) -> Unit)? = null
    private var pendingDocumentCallback: ((Result<DocumentScanResult?>) -> Unit)? = null
    private var pendingGalleryCallback: ((Result<BarcodeResult?>) -> Unit)? = null
    private var pendingGalleryFormats: List<String>? = null

    // Document scanner components
    private var documentScanner: GmsDocumentScanner? = null
    private var scannerLauncher: ActivityResultLauncher<IntentSenderRequest>? = null

    // Torch state tracking
    private var torchEnabled = false

    /** Reference to active embedded scanner for runtime configuration updates */
    internal var activeEmbeddedScanner: EmbeddedScannerView? = null

    // MARK: - FlutterPlugin Implementation

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        flutterApi = ScannerFlutterApi(flutterPluginBinding.binaryMessenger)

        // Set up Pigeon API
        ScannerHostApi.setUp(flutterPluginBinding.binaryMessenger, this)

        // Register platform view factory for embedded scanner
        flutterPluginBinding.platformViewRegistry.registerViewFactory(
            "dev.scankit/scanner",
            ScannerViewFactory(flutterPluginBinding.binaryMessenger, this)
        )
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        ScannerHostApi.setUp(binding.binaryMessenger, null)
        flutterApi = null
        context = null
    }

    // MARK: - ActivityAware Implementation

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        activityBinding = binding

        // Register activity result listener
        binding.addActivityResultListener { requestCode, resultCode, data ->
            handleActivityResult(requestCode, resultCode, data)
        }

        // Register document scanner launcher if activity supports it
        (binding.activity as? AppCompatActivity)?.let { appCompatActivity ->
            scannerLauncher = appCompatActivity.registerForActivityResult(
                ActivityResultContracts.StartIntentSenderForResult()
            ) { result ->
                handleDocumentScanResult(result.resultCode, result.data)
            }
        }
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
        activityBinding = null
        scannerLauncher = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
        activityBinding = binding

        (binding.activity as? AppCompatActivity)?.let { appCompatActivity ->
            scannerLauncher = appCompatActivity.registerForActivityResult(
                ActivityResultContracts.StartIntentSenderForResult()
            ) { result ->
                handleDocumentScanResult(result.resultCode, result.data)
            }
        }
    }

    override fun onDetachedFromActivity() {
        activity = null
        activityBinding = null
        scannerLauncher = null
    }

    // MARK: - Device Capabilities

    override fun isSupported(): Boolean = true

    override fun isTorchAvailable(): Boolean {
        val ctx = context ?: return false
        val cameraManager = ctx.getSystemService(Context.CAMERA_SERVICE) as? CameraManager
            ?: return false

        return try {
            val cameraId = cameraManager.cameraIdList.firstOrNull() ?: return false
            val characteristics = cameraManager.getCameraCharacteristics(cameraId)
            characteristics.get(CameraCharacteristics.FLASH_INFO_AVAILABLE) == true
        } catch (e: Exception) {
            false
        }
    }

    override fun isDocumentScanSupported(): Boolean = true

    // MARK: - Torch Control

    override fun setTorch(enabled: Boolean) {
        torchEnabled = enabled
        flutterApi?.onTorchStateChanged(enabled) { }
    }

    override fun getTorchState(): Boolean = torchEnabled

    // MARK: - Full-Screen Barcode Scanning

    override fun startScan(config: ScannerConfig, callback: (Result<BarcodeResult?>) -> Unit) {
        val currentActivity = activity
        if (currentActivity == null) {
            callback(Result.failure(FlutterError("NO_ACTIVITY", "No activity available", null)))
            return
        }

        pendingCallback = callback

        val intent = Intent(currentActivity, ScannerActivity::class.java).apply {
            putStringArrayListExtra(ScannerActivity.EXTRA_FORMATS, config.formats?.let { ArrayList(it) })
            putExtra(ScannerActivity.EXTRA_VIBRATE, config.vibrateOnScan)
            putExtra(ScannerActivity.EXTRA_BEEP, config.beepOnScan)
            putExtra(ScannerActivity.EXTRA_SHOW_TORCH, config.showTorchButton)
        }

        currentActivity.startActivityForResult(intent, REQUEST_CODE_SCAN)
    }

    // MARK: - Document Scanning

    override fun startDocumentScan(config: DocumentScanConfig, callback: (Result<DocumentScanResult?>) -> Unit) {
        val currentActivity = activity
        if (currentActivity == null) {
            callback(Result.failure(FlutterError("NO_ACTIVITY", "No activity available", null)))
            return
        }

        pendingDocumentCallback = callback

        val optionsBuilder = GmsDocumentScannerOptions.Builder()
            .setGalleryImportAllowed(config.allowGalleryImport)
            .setResultFormats(GmsDocumentScannerOptions.RESULT_FORMAT_JPEG)
            .setScannerMode(GmsDocumentScannerOptions.SCANNER_MODE_FULL)

        config.maxPages?.let { maxPages ->
            optionsBuilder.setPageLimit(maxPages.toInt())
        }

        documentScanner = GmsDocumentScanning.getClient(optionsBuilder.build())

        documentScanner?.getStartScanIntent(currentActivity)
            ?.addOnSuccessListener { intentSender ->
                val launcher = scannerLauncher
                if (launcher != null) {
                    launcher.launch(IntentSenderRequest.Builder(intentSender).build())
                } else {
                    // Fallback for non-AppCompatActivity
                    currentActivity.startIntentSenderForResult(
                        intentSender,
                        REQUEST_CODE_DOCUMENT_SCAN,
                        null, 0, 0, 0
                    )
                }
            }
            ?.addOnFailureListener { e ->
                pendingDocumentCallback?.invoke(Result.failure(FlutterError(
                    "DOCUMENT_SCAN_ERROR",
                    e.message ?: "Failed to start scanner",
                    null
                )))
                pendingDocumentCallback = null
            }
    }

    // MARK: - Image Scanning

    override fun scanFromImagePath(imagePath: String, formats: List<String>?, callback: (Result<BarcodeResult?>) -> Unit) {
        ImageScanner.scanFromFile(imagePath, formats, callback)
    }

    override fun scanFromGallery(formats: List<String>?, callback: (Result<BarcodeResult?>) -> Unit) {
        val currentActivity = activity
        if (currentActivity == null) {
            callback(Result.failure(FlutterError("NO_ACTIVITY", "No activity available", null)))
            return
        }

        pendingGalleryCallback = callback
        pendingGalleryFormats = formats

        val intent = Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI)
        currentActivity.startActivityForResult(intent, REQUEST_CODE_GALLERY)
    }

    // MARK: - Warm-up

    override fun warmUp() {
        // Pre-initialize ML Kit barcode scanner for faster first scan
        val options = BarcodeScannerOptions.Builder()
            .setBarcodeFormats(Barcode.FORMAT_ALL_FORMATS)
            .build()
        BarcodeScanning.getClient(options)
    }

    // MARK: - Embedded Scanner Configuration

    override fun updateEmbeddedConfig(config: EmbeddedScannerConfig) {
        activeEmbeddedScanner?.updateConfig(config)
    }

    override fun switchCamera() {
        activeEmbeddedScanner?.switchCamera()
    }

    override fun setZoom(zoom: Double) {
        activeEmbeddedScanner?.setZoom(zoom.toFloat())
    }

    // MARK: - Flutter Callbacks

    /** Notify Flutter of a detected barcode */
    fun notifyBarcodeDetected(result: BarcodeResult) {
        flutterApi?.onBarcodeDetected(result) { }
    }

    /** Notify Flutter of an error */
    fun notifyError(code: String, message: String) {
        flutterApi?.onError(code, message) { }
    }

    /** Notify Flutter of a camera switch */
    fun notifyCameraSwitched(facing: CameraFacing) {
        flutterApi?.onCameraSwitched(facing) { }
    }

    /** Notify Flutter of a zoom change */
    fun notifyZoomChanged(zoom: Double) {
        flutterApi?.onZoomChanged(zoom) { }
    }

    // MARK: - Activity Result Handling

    private fun handleActivityResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean {
        when (requestCode) {
            REQUEST_CODE_SCAN -> {
                handleScanResult(resultCode, data)
                return true
            }
            REQUEST_CODE_DOCUMENT_SCAN -> {
                handleDocumentScanResult(resultCode, data)
                return true
            }
            REQUEST_CODE_GALLERY -> {
                handleGalleryResult(resultCode, data)
                return true
            }
        }
        return false
    }

    private fun handleScanResult(resultCode: Int, data: Intent?) {
        val callback = pendingCallback
        pendingCallback = null

        when (resultCode) {
            Activity.RESULT_OK -> {
                val value = data?.getStringExtra(ScannerActivity.RESULT_VALUE)
                val format = data?.getStringExtra(ScannerActivity.RESULT_FORMAT)
                val rawBytes = data?.getByteArrayExtra(ScannerActivity.RESULT_RAW_BYTES)

                // Get bounding box
                val boxLeft = data?.getDoubleExtra(ScannerActivity.RESULT_BOX_LEFT, -1.0) ?: -1.0
                val boxTop = data?.getDoubleExtra(ScannerActivity.RESULT_BOX_TOP, -1.0) ?: -1.0
                val boxRight = data?.getDoubleExtra(ScannerActivity.RESULT_BOX_RIGHT, -1.0) ?: -1.0
                val boxBottom = data?.getDoubleExtra(ScannerActivity.RESULT_BOX_BOTTOM, -1.0) ?: -1.0

                val boundingBox = if (boxLeft >= 0) {
                    BoundingBox(boxLeft, boxTop, boxRight, boxBottom)
                } else null

                // Get corner points
                val cornerPointsArray = data?.getDoubleArrayExtra(ScannerActivity.RESULT_CORNERS)
                val cornerPoints = cornerPointsArray?.let { arr ->
                    if (arr.size >= 8) {
                        listOf(
                            Point(arr[0], arr[1]),
                            Point(arr[2], arr[3]),
                            Point(arr[4], arr[5]),
                            Point(arr[6], arr[7])
                        )
                    } else null
                }

                if (value != null && format != null) {
                    callback?.invoke(Result.success(BarcodeResult(
                        value, format, rawBytes, boundingBox, cornerPoints
                    )))
                } else {
                    callback?.invoke(Result.success(null))
                }
            }
            Activity.RESULT_CANCELED -> {
                callback?.invoke(Result.success(null))
            }
            else -> {
                val error = data?.getStringExtra(ScannerActivity.RESULT_ERROR)
                callback?.invoke(Result.failure(FlutterError(
                    "SCAN_ERROR",
                    error ?: "Unknown error",
                    null
                )))
            }
        }
    }

    private fun handleDocumentScanResult(resultCode: Int, data: Intent?) {
        val callback = pendingDocumentCallback
        pendingDocumentCallback = null

        if (resultCode == Activity.RESULT_OK && data != null) {
            val scanResult = GmsDocumentScanningResult.fromActivityResultIntent(data)
            if (scanResult != null) {
                val pages = scanResult.pages?.map { page ->
                    ScannedPage(
                        imagePath = page.imageUri.path ?: "",
                        width = 0,
                        height = 0
                    )
                } ?: emptyList()

                val result = DocumentScanResult(
                    pages = pages,
                    pdfPath = scanResult.pdf?.uri?.path
                )
                callback?.invoke(Result.success(result))
            } else {
                callback?.invoke(Result.success(null))
            }
        } else if (resultCode == Activity.RESULT_CANCELED) {
            callback?.invoke(Result.success(null))
        } else {
            callback?.invoke(Result.failure(FlutterError(
                "DOCUMENT_SCAN_ERROR",
                "Document scan failed",
                null
            )))
        }
    }

    private fun handleGalleryResult(resultCode: Int, data: Intent?) {
        val callback = pendingGalleryCallback
        val formats = pendingGalleryFormats
        pendingGalleryCallback = null
        pendingGalleryFormats = null

        if (resultCode != Activity.RESULT_OK || data?.data == null) {
            callback?.invoke(Result.success(null))
            return
        }

        val ctx = context
        if (ctx == null) {
            callback?.invoke(Result.failure(FlutterError("NO_CONTEXT", "No context available", null)))
            return
        }

        try {
            val uri = data.data!!
            val image = InputImage.fromFilePath(ctx, uri)
            ImageScanner.scanFromImage(image, formats, image.width, image.height, callback!!)
        } catch (e: Exception) {
            callback?.invoke(Result.failure(FlutterError(
                "IMAGE_ERROR",
                e.message ?: "Failed to load image",
                null
            )))
        }
    }

    companion object {
        private const val REQUEST_CODE_SCAN = 29438
        private const val REQUEST_CODE_DOCUMENT_SCAN = 29439
        private const val REQUEST_CODE_GALLERY = 29440
    }
}

/**
 * Factory for creating embedded scanner platform views.
 * Registered with Flutter to handle "dev.scankit/scanner" view type.
 */
class ScannerViewFactory(
    private val messenger: BinaryMessenger,
    private val plugin: ScankitPlugin
) : PlatformViewFactory(StandardMessageCodec.INSTANCE) {

    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        val params = args as? Map<*, *>
        val formats = (params?.get("formats") as? List<*>)?.filterIsInstance<String>()
        return EmbeddedScannerView(context, viewId, formats, plugin)
    }
}

File: ios/Classes/ScankitPlugin.swift
// ScankitPlugin.swift
// ScanKit - Modern barcode scanner for Flutter
//
// Main plugin entry point that implements the Pigeon-generated ScannerHostApi.
// This file coordinates between Flutter and native scanner implementations.
//
// Architecture:
// - ScankitPlugin: Main plugin, handles Pigeon API calls
// - EmbeddedScannerView: Platform view for embedded scanning (Scanner/)
// - FullScreenScannerPresenter: Full-screen scanner UI (Scanner/)
// - DocumentScannerPresenter: Document scanning (Scanner/)
// - ImageScanner/GalleryImagePicker: Static image scanning (Scanner/)
// - BarcodeFormatMapper: Format string <-> symbology mapping (Utils/)
// - TorchController: Torch/flashlight control (Utils/)

import Flutter
import UIKit
import VisionKit
import AVFoundation
import Vision

// MARK: - Main Plugin

@MainActor
public class ScankitPlugin: NSObject, FlutterPlugin, ScannerHostApi {

    // MARK: - Properties

    /// Pigeon API for callbacks to Flutter
    private var flutterApi: ScannerFlutterApi?

    /// Plugin registrar reference
    private weak var registrar: FlutterPluginRegistrar?

    /// Active embedded scanner (for runtime config updates)
    weak var activeEmbeddedScanner: EmbeddedScannerView?

    /// Full-screen scanner presenter
    private var fullScreenPresenter: FullScreenScannerPresenter?

    /// Document scanner presenter
    private var documentPresenter: DocumentScannerPresenter?

    /// Gallery image picker
    private var galleryPicker: GalleryImagePicker?

    /// Pending results for async operations
    private var pendingResult: ((Result<BarcodeResult?, Error>) -> Void)?
    private var pendingDocumentResult: ((Result<DocumentScanResult?, Error>) -> Void)?

    // MARK: - Plugin Registration

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = ScankitPlugin()
        instance.registrar = registrar
        instance.flutterApi = ScannerFlutterApi(binaryMessenger: registrar.messenger())

        // Set up Pigeon API
        ScannerHostApiSetup.setUp(binaryMessenger: registrar.messenger(), api: instance)

        // Register platform view factory for embedded scanner
        let factory = ScannerViewFactory(messenger: registrar.messenger(), plugin: instance)
        registrar.register(factory, withId: "dev.scankit/scanner")
    }

    // MARK: - Device Capabilities

    nonisolated func isSupported() throws -> Bool {
        if #available(iOS 16.0, *) {
            return MainActor.assumeIsolated {
                DataScannerViewController.isSupported && DataScannerViewController.isAvailable
            }
        }
        return AVCaptureDevice.default(for: .video) != nil
    }

    nonisolated func isTorchAvailable() throws -> Bool {
        return TorchController.isTorchAvailable()
    }

    nonisolated func isDocumentScanSupported() throws -> Bool {
        return DocumentScannerPresenter.isSupported
    }

    // MARK: - Torch Control

    nonisolated func setTorch(enabled: Bool) throws {
        let success = TorchController.setTorch(enabled: enabled)
        if success {
            Task { @MainActor in
                self.flutterApi?.onTorchStateChanged(enabled: enabled) { _ in }
            }
        }
    }

    nonisolated func getTorchState() throws -> Bool {
        return TorchController.getTorchState()
    }

    // MARK: - Full-Screen Barcode Scanning

    nonisolated func startScan(config: ScannerConfig, completion: @escaping (Result<BarcodeResult?, Error>) -> Void) {
        Task { @MainActor in
            self.pendingResult = completion

            if self.fullScreenPresenter == nil {
                self.fullScreenPresenter = FullScreenScannerPresenter(plugin: self)
            }

            self.fullScreenPresenter?.presentScanner(config: config) { result in
                self.pendingResult = nil
                completion(result)
            }
        }
    }

    // MARK: - Document Scanning

    nonisolated func startDocumentScan(config: DocumentScanConfig, completion: @escaping (Result<DocumentScanResult?, Error>) -> Void) {
        Task { @MainActor in
            self.pendingDocumentResult = completion

            if self.documentPresenter == nil {
                self.documentPresenter = DocumentScannerPresenter()
            }

            self.documentPresenter?.presentDocumentScanner(config: config) { result in
                self.pendingDocumentResult = nil
                completion(result)
            }
        }
    }

    // MARK: - Image Scanning

    nonisolated func scanFromImagePath(imagePath: String, formats: [String]?, completion: @escaping (Result<BarcodeResult?, Error>) -> Void) {
        Task { @MainActor in
            ImageScanner.scanBarcode(fromImagePath: imagePath, formats: formats, completion: completion)
        }
    }

    nonisolated func scanFromGallery(formats: [String]?, completion: @escaping (Result<BarcodeResult?, Error>) -> Void) {
        Task { @MainActor in
            if self.galleryPicker == nil {
                self.galleryPicker = GalleryImagePicker()
            }

            self.galleryPicker?.pickAndScanImage(formats: formats, completion: completion)
        }
    }

    // MARK: - Warm-up

    nonisolated func warmUp() throws {
        // Pre-initialize Vision framework for faster first scan
        Task { @MainActor in
            let request = VNDetectBarcodesRequest { _, _ in }
            if #available(iOS 15.0, *) {
                _ = request.supportedSymbologies
            }
        }
    }

    // MARK: - Embedded Scanner Configuration

    nonisolated func updateEmbeddedConfig(config: EmbeddedScannerConfig) throws {
        Task { @MainActor in
            self.activeEmbeddedScanner?.updateConfig(config)
        }
    }

    nonisolated func switchCamera() throws {
        Task { @MainActor in
            self.activeEmbeddedScanner?.switchCamera()
        }
    }

    nonisolated func setZoom(zoom: Double) throws {
        Task { @MainActor in
            self.activeEmbeddedScanner?.setZoom(CGFloat(zoom))
        }
    }

    // MARK: - Flutter Callbacks

    /// Notify Flutter of a detected barcode.
    func notifyBarcodeDetected(_ result: BarcodeResult) {
        flutterApi?.onBarcodeDetected(barcode: result) { _ in }
    }

    /// Notify Flutter of a camera switch.
    func notifyCameraSwitched(_ facing: CameraFacing) {
        flutterApi?.onCameraSwitched(facing: facing) { _ in }
    }

    /// Notify Flutter of a zoom change.
    func notifyZoomChanged(_ zoom: Double) {
        flutterApi?.onZoomChanged(zoom: zoom) { _ in }
    }

    /// Notify Flutter of an error.
    func notifyError(code: String, message: String) {
        flutterApi?.onError(code: code, message: message) { _ in }
    }
}

// MARK: - DataScannerViewControllerDelegate

@available(iOS 16.0, *)
extension ScankitPlugin: DataScannerViewControllerDelegate {

    public func dataScanner(_ dataScanner: DataScannerViewController, didTapOn item: RecognizedItem) {
        processRecognizedItem(item, viewSize: dataScanner.view.bounds.size)
    }

    public func dataScanner(_ dataScanner: DataScannerViewController, didAdd addedItems: [RecognizedItem], allItems: [RecognizedItem]) {
        if let firstItem = addedItems.first {
            processRecognizedItem(firstItem, viewSize: dataScanner.view.bounds.size)
        }
    }

    public func dataScannerDidZoom(_ dataScanner: DataScannerViewController) {}

    public func dataScanner(_ dataScanner: DataScannerViewController, becameUnavailableWithError error: DataScannerViewController.ScanningUnavailable) {
        fullScreenPresenter?.handleError(PigeonError(
            code: "SCANNER_ERROR",
            message: error.localizedDescription,
            details: nil
        ))
    }

    private func processRecognizedItem(_ item: RecognizedItem, viewSize: CGSize) {
        guard case .barcode(let barcode) = item else { return }

        let format = BarcodeFormatMapper.symbologyToFormat(barcode.observation.symbology)
        let value = barcode.payloadStringValue ?? ""

        // Calculate bounding box (normalized)
        let bounds = barcode.bounds
        let minX = min(bounds.topLeft.x, bounds.bottomLeft.x)
        let maxX = max(bounds.topRight.x, bounds.bottomRight.x)
        let minY = min(bounds.topLeft.y, bounds.topRight.y)
        let maxY = max(bounds.bottomLeft.y, bounds.bottomRight.y)

        let boundingBox = BoundingBox(
            left: minX / viewSize.width,
            top: minY / viewSize.height,
            right: maxX / viewSize.width,
            bottom: maxY / viewSize.height
        )

        // Corner points (normalized)
        let cornerPoints: [Point] = [
            Point(x: Double(bounds.topLeft.x / viewSize.width), y: Double(bounds.topLeft.y / viewSize.height)),
            Point(x: Double(bounds.topRight.x / viewSize.width), y: Double(bounds.topRight.y / viewSize.height)),
            Point(x: Double(bounds.bottomRight.x / viewSize.width), y: Double(bounds.bottomRight.y / viewSize.height)),
            Point(x: Double(bounds.bottomLeft.x / viewSize.width), y: Double(bounds.bottomLeft.y / viewSize.height)),
        ]

        let result = BarcodeResult(
            value: value,
            format: format,
            rawBytes: nil,
            boundingBox: boundingBox,
            cornerPoints: cornerPoints
        )

        fullScreenPresenter?.handleScanResult(result)
    }
}

// MARK: - UIAdaptivePresentationControllerDelegate

extension ScankitPlugin: UIAdaptivePresentationControllerDelegate {

    public func presentationControllerDidDismiss(_ presentationController: UIPresentationController) {
        // User dismissed by swiping down
        fullScreenPresenter?.handleCancel()
    }
}
